storage/transfermanager/download_buffer.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transfermanager
+
+import (
+	"sync"
+)
+
+// NewDownloadBuffer initializes a DownloadBuffer using buf as the underlying
+// buffer. Preferred way to create a DownloadBuffer as it does not need to grow
+// the buffer if len(buf) is larger than or equal to the object length or range
+// being downloaded to.
+func NewDownloadBuffer(buf []byte) *DownloadBuffer {
+	return &DownloadBuffer{bytes: buf}
+}
+
+// DownloadBuffer satisfies the io.WriterAt interface, allowing you to use it as
+// a buffer to download to when using [Downloader]. DownloadBuffer is thread-safe
+// as long as the ranges being written to do not overlap.
+type DownloadBuffer struct {
+	bytes []byte
+	mu    sync.Mutex
+}
+
+// WriteAt writes len(p) bytes from p to the underlying buffer at offset off,
+// growing the buffer if needed. It returns the number of bytes written from p
+// and any error encountered that caused the write to stop early.
+// WriteAt is thread-safe as long as the ranges being written to do not overlap.
+// The supplied slice p is not retained.
+func (db *DownloadBuffer) WriteAt(p []byte, off int64) (n int, err error) {
+	requiredLength := int64(len(p)) + off
+
+	// Our buffer isn't big enough, let's grow it.
+	if int64(cap(db.bytes)) < requiredLength {
```

Comment:
It's completely possible that that happens (depends on partsize), but I think we also don't want to increment by a lot more than is needed. I was initially thinking to double it as `append` does, but that could more easily run into problems with large buffers.

To avoid a lot of small increments, users could pass in a buffer that's already close to the object size to NewDownloadBuffer, or we could schedule the last shard first, so that it grows the buffer all it needs to since the beginning (but that has other potential performance implications).  Another option is to give users the ability to configure how much the buffer grows by.

Reply:
I think even if partsize is large it still might be possible for many small writes to happen. But yeah profiling is the way to diagnose whether this is an issue.

I think we can add the growth rate later as an exported field, so we can merge as-is and just add that later if need be.